zipline/data/adjustments.py

```diff
@@ -349,72 +350,62 @@ def calc_dividend_ratios(self, dividends):
                 [],
                 dtype=[
-                    ('sid', uint32_dtype),
+                    ('sid', uint64_dtype),
                     ('effective_date', uint32_dtype),
                     ('ratio', float64_dtype),
                 ],
             ))
-        ex_dates = dividends.ex_date.values
-
-        sids = dividends.sid.values
-        amounts = dividends.amount.values
-
-        ratios = np.full(len(amounts), np.nan)
-
-        equity_daily_bar_reader = self._equity_daily_bar_reader
-
-        effective_dates = np.full(len(amounts), -1, dtype=int64_dtype)
-
-        calendar = self._calendar
-
-        # Calculate locs against a tz-naive cal, as the ex_dates are tz-
-        # naive.
-        #
-        # TODO: A better approach here would be to localize ex_date to
-        # the tz of the calendar, but currently get_indexer does not
-        # preserve tz of the target when method='bfill', which throws
-        # off the comparison.
-        tz_naive_calendar = calendar.tz_localize(None)
-        day_locs = tz_naive_calendar.get_indexer(ex_dates, method='bfill')
-
-        isnull = pd.isnull
-
-        for i, amount in enumerate(amounts):
-            sid = sids[i]
-            ex_date = ex_dates[i]
-            day_loc = day_locs[i]
-
-            prev_close_date = calendar[day_loc - 1]
-
-            try:
-                prev_close = equity_daily_bar_reader.get_value(
-                    sid, prev_close_date, 'close')
-                if not isnull(prev_close):
-                    ratio = 1.0 - amount / prev_close
-                    ratios[i] = ratio
-                    # only assign effective_date when data is found
-                    effective_dates[i] = ex_date
-            except NoDataOnDate:
-                log.warn("Couldn't compute ratio for dividend %s" % {
-                    'sid': sid,
-                    'ex_date': ex_date,
-                    'amount': amount,
-                })
-                continue
-
-        # Create a mask to filter out indices in the effective_date, sid, and
-        # ratio vectors for which a ratio was not calculable.
-        effective_mask = effective_dates != -1
-        effective_dates = effective_dates[effective_mask]
-        effective_dates = effective_dates.astype('datetime64[ns]').\
-            astype('datetime64[s]').astype(uint32_dtype)
-        sids = sids[effective_mask]
-        ratios = ratios[effective_mask]
 
+        pricing_reader = self._equity_daily_bar_reader
+        input_sids = dividends.sid.values
+        unique_sids, sids_ix = np.unique(input_sids, return_inverse=True)
+        dates = pricing_reader.sessions.values
+
+        close, = pricing_reader.load_raw_arrays(
+            ['close'],
+            pd.Timestamp(dates[0], tz='UTC'),
+            pd.Timestamp(dates[-1], tz='UTC'),
+            unique_sids,
+        )
+        date_ix = np.searchsorted(dates, dividends.ex_date.values)
+        mask = date_ix > 0
```

Comment:
I don't think this can ever be False.

```python
In [6]: data = np.arange(5)

In [7]: data.searchsorted([-1, 3, 6])
Out[7]: array([0, 3, 5])
```

I think ^ also means we need to handle the case of dates that are outside the range or misaligned with the input index. The way I've written that lookup before is:

```python
def loc_as_of(dts, asof):
    """
    Find the position of the last entry in ``dts`` on or before ``as_of``.

    Parameters
    ----------
    dts : ndarray or DatetimeIndex
        Array of sorted datetime-like objects.
    as_of : pd.Timestamp
        Timestamp to find the position as of.
    """
    ix = dts.searchsorted(asof.asm8, side='right') - 1
    if ix < 0:
        raise IndexError("{} is before first value {}", asof, dts[0])
    return ix
```

I think this algorithm vectorizes straightforwardly, though I'm not sure the way it handles boundaries is what we want. If we get an `ex_date` in the future or too far in the past, what do we want to do? My guess is we probably just shouldn't compute a ratio for them.

Reply:
I am not sure what your first example is showing, the first element of that would be false:
```
In [9]: data = np.arange(5)

In [10]: ix = data.searchsorted([-1, 3, 6])

In [11]: ix > 0
Out[11]: array([False,  True,  True], dtype=bool)
```

Reply:
Ah, yeah, sorry, I read it as `>=`.